rmw_iceoryx_cpp/src/rmw_wait.cpp

```diff
@@ -49,107 +51,78 @@ rmw_wait(
-    if (nsec >= 1000000000L) {
-      nsec -= 1000000000L;
-      ts_end.tv_sec = ts_start.tv_sec + wait_timeout->sec + 1;
-    } else {
-      ts_end.tv_sec = ts_start.tv_sec + wait_timeout->sec;
-    }
-    ts_end.tv_nsec = nsec;
-    semaphore->timedWait(&ts_end, true);
-
-    // // for debugging
-    // struct timespec ts_diff;
-    // clock_gettime(CLOCK_REALTIME, &ts_diff);
-    // int64_t diff_nsec = ts_diff.tv_nsec - ts_start.tv_nsec;
-    // if (diff_nsec < 0L) {
-    //   diff_nsec += 1000000000L;
-    //   ts_diff.tv_sec = ts_diff.tv_sec - ts_start.tv_sec - 1;
-    // } else {
-    //   ts_diff.tv_sec = ts_diff.tv_sec - ts_start.tv_sec;
-    // }
-    // ts_diff.tv_nsec = diff_nsec;
-    // printf("waited s %lu ns %lu\n", ts_diff.tv_sec, ts_diff.tv_nsec);
+    auto sec = iox::units::Duration::fromSeconds(wait_timeout->sec);
+    auto nsec = iox::units::Duration::fromNanoseconds(wait_timeout->nsec);
+    auto timeout = sec + nsec;
+
+    /// @todo Check triggered subscribers in vector? Is that relevant for rmw?
+    auto notificationVector = waitset->timedWait(iox::units::Duration(timeout));
   }
 
 after_wait:
-
-
   // reset all the subscriptions that don't have new data
   for (size_t i = 0; i < subscriptions->subscriber_count; ++i) {
     auto iceoryx_subscription =
-      static_cast<IceoryxSubscription * const>(subscriptions->subscribers[i]);
-    iox::popo::Subscriber * iceoryx_receiver = iceoryx_subscription->iceoryx_receiver_;
+      static_cast<IceoryxSubscription *>(subscriptions->subscribers[i]);
+    iox::popo::UntypedSubscriber * iceoryx_receiver = iceoryx_subscription->iceoryx_receiver_;
 
-    // remove semaphore from all receivers because next call a new waitset could be provided
-    iceoryx_receiver->unsetChunkReceiveSemaphore();
+    // remove waitset from all receivers because next call a new waitset could be provided
+    waitset->detachState(*iceoryx_receiver, iox::popo::SubscriberState::HAS_DATA);
 
-    if (!iceoryx_receiver->hasNewChunks()) {
+    if (!iceoryx_receiver->hasData()) {
       subscriptions->subscribers[i] = nullptr;
     }
   }
 
   // reset all the guard_conditions that have not triggered
   for (size_t i = 0; i < guard_conditions->guard_condition_count; ++i) {
     auto iceoryx_guard_condition =
-      static_cast<IceoryxGuardCondition * const>(guard_conditions->guard_conditions[i]);
-
-    iceoryx_guard_condition->detachSemaphore();
+      static_cast<iox::popo::UserTrigger *>(guard_conditions->guard_conditions[i]);
 
-    if (iceoryx_guard_condition->hasTriggered()) {
-      iceoryx_guard_condition->resetTriggerIndication();
-    } else {
-      guard_conditions->guard_conditions[i] = nullptr;
-    }
-  }
+    waitset->detachEvent(*iceoryx_guard_condition);
 
-  // clear the semaphore
-  // events that triggered it and where not yet collected will be seen on next rmw_wait
-  while (semaphore->tryWait()) {
+    guard_conditions->guard_conditions[i] = nullptr;
```

Comment:
Only set the guard conditions to nullptr which haven't triggered

Reply:
```
    if (!iceoryx_guard_condition->hasTriggered()) {
      guard_conditions->guard_conditions[i] = nullptr;
    }
```

Reply:
done.